protocols/relay/src/v2/relay.rs

```diff
@@ -0,0 +1,643 @@
+// Copyright 2021 Protocol Labs.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+//! [`NetworkBehaviour`] to act as a circuit relay v2 **relay**.
+
+mod handler;
+pub mod rate_limiter;
+
+use crate::v2::message_proto;
+use libp2p_core::connection::{ConnectedPoint, ConnectionId};
+use libp2p_core::multiaddr::Protocol;
+use libp2p_core::{Multiaddr, PeerId};
+use libp2p_swarm::{NetworkBehaviour, NetworkBehaviourAction, NotifyHandler, PollParameters};
+use std::collections::{HashMap, HashSet, VecDeque};
+use std::num::NonZeroU32;
+use std::ops::Add;
+use std::task::{Context, Poll};
+use std::time::Duration;
+use wasm_timer::Instant;
+
+/// Configuration for the [`Relay`] [`NetworkBehaviour`].
+///
+/// # Panics
+///
+/// [`Config::max_circuit_duration`] may not exceed [`u32::MAX`].
+pub struct Config {
```

Comment:
Please implement debug for config structs.

Reply:
Continued in https://github.com/libp2p/rust-libp2p/pull/2059/files#r678447392.